bullet/src/EntityManagementFeatures.cc

```diff
@@ -0,0 +1,213 @@
+{
+  // Create bullet empty multibody dynamics world
+  const auto collisionConfiguration = std::make_shared<btDefaultCollisionConfiguration>();
+  const auto dispatcher =
+    std::make_shared<btCollisionDispatcher>(collisionConfiguration.get());
+  //const auto broadphase = std::shared_ptr<btBroadphaseInterface>(new btDbvtBroadphase());
+  const auto broadphase = std::make_shared<btDbvtBroadphase>();
+  const auto solver =
+    std::make_shared<btSequentialImpulseConstraintSolver>();
+  const auto world = std::make_shared<btDiscreteDynamicsWorld>(
+    dispatcher.get(), broadphase.get(), solver.get(), collisionConfiguration.get());
+
+  /* TO-DO(Lobotuerk): figure out what this line does*/
+  world->getSolverInfo().m_globalCfm = 0;
+
+  btGImpactCollisionAlgorithm::registerAlgorithm(dispatcher.get());
+
+  return this->AddWorld(
+    {_name, collisionConfiguration, dispatcher, broadphase, solver, world});
+}
+
+/////////////////////////////////////////////////
+bool EntityManagementFeatures::RemoveModel(const Identity &_modelID)
+{
+  // Check if the model exists
+  if (this->models.find(_modelID.id) == this->models.end())
+  {
+    return false;
+  }
+
+  auto model = this->models.at(_modelID);
+  auto bulletWorld = this->worlds.at(model->world)->world;
+
+  // Clean up joints, this section considers both links in the joint
+  // are part of the same world
+  auto joint_it = this->joints.begin();
+  while (joint_it != this->joints.end())
+  {
+    const auto &jointInfo = joint_it->second;
+    const auto &childLinkInfo = this->links[jointInfo->childLinkId];
+    if (childLinkInfo->model.id == _modelID.id)
+    {
+      bulletWorld->removeConstraint(jointInfo->joint.get());
+      joint_it = this->joints.erase(joint_it);
+      continue;
+    }
+    joint_it++;
+  }
+
+  // Clean up collisions
+  auto collision_it = this->collisions.begin();
+  while (collision_it != this->collisions.end())
+  {
+    const auto &collisionInfo = collision_it->second;
+    if (collisionInfo->model.id == _modelID.id)
+    {
+      collision_it = this->collisions.erase(collision_it);
+      continue;
+    }
+    collision_it++;
+  }
+
+  // Clean up links
+  auto it = this->links.begin();
+  while (it != this->links.end())
+  {
+    const auto &linkInfo = it->second;
+    if (linkInfo->model.id == _modelID.id)
+    {
+      bulletWorld->removeRigidBody(linkInfo->link.get());
+      it = this->links.erase(it);
+      continue;
+    }
+    it++;
+  }
+
+  // Clean up model, links are erased and the model is just a name to tie
+  // them together
+  this->models.erase(_modelID.id);
+
+  return true;
+}
+
+bool EntityManagementFeatures::ModelRemoved(
+    const Identity &_modelID) const
+{
+  return this->models.find(_modelID) == this->models.end();
+}
+
+bool EntityManagementFeatures::RemoveModelByIndex(
+    const Identity & _worldID, std::size_t _modelIndex)
+{
+  // Check if the model exists
+  if (this->models.find(_modelIndex) == this->models.end() ||
```

Comment:
Wouldn't this check against model IDs instead of indices?

Reply:
Yeah, we are using the generated entities values as indices to save the objects inside unordered maps. So, there is an equivalence modelID.id to index. Would this cause a problem? Maybe we are missing something

Reply:
Well, it depends on the way the IDs are generated. Right now, the ID is incremented for any entity created. Consider this SDF
```xml
<sdf version='1.6'>
  <world name='W'>
    <model name='M0'>
      <link name='L'/>
    </model>
   <model name='M1'/>
  </world>
</sdf>
```

The index of model `M1` in the world should always be 1, but if the order of creation happens to be `W`, `M0`, `L`, `M1`, the ID of `M1` would be 3!

This is also problematic when models are removed and created at run time. The ID would always increase, but the index may not.

Reply:
>> The index of model M1 in the world should always be 1

Why? Is this information saved/generated elsewhere? Like something checking how many models are in each world?

Reply:
Yeah, if you want to loop through all the models, you can do a for loop starting from 0 to `World::ModelCount()`.

Reply:
Should be addressed with #240, PTAL